node/beacon/src/authority.rs

```diff
@@ -57,12 +61,17 @@ pub struct Authority {
     participation: HashMap<Slot, AuthorityMask>,
```

Comment:
This implementation should be doing caching in the same way we do caching for `storage/src/storages/{beacon.rs, shard.rs}`. Specifically, collections like `thresholds`, `accepted_authorities`, `processed_blocks`, `participation`, `proposals` should be used similarly to how we use e.g. `transaction_results` in `ShardChainStorage`. In other words, they should be moved into `storage/src/storages/beacon.rs`.

We then should implement pruning as the generic feature of our `storage/src/storages/mod.rs` because we will be applying it to other columns and not just authority-related data.

Reply:
The API for reading from and writing to storage is very limited and does not work well with the code in authority that uses them, unless we expose more methods like iterators, entry, etc.

Reply:
Why do we need to expose iterators and entry?

Reply:
Also, the other point that I brought is that the implementation should be explicit about what containers are used as cache, and these cache containers should live in storage structs and not authorities.

Reply:
Also, we want pruning to be a feature of the storage, so that we can apply it to other stuff and not just authorities. And it will prevent code duplication, because otherwise every component that uses storage will be implementing its own version of pruning.

Reply:
There are things like https://github.com/nearprotocol/nearcore/blob/7ffb2698fc9fe4b908423f13b9352856cdaab26a/node/beacon/src/authority.rs#L281, https://github.com/nearprotocol/nearcore/blob/7ffb2698fc9fe4b908423f13b9352856cdaab26a/node/beacon/src/authority.rs#L273,
https://github.com/nearprotocol/nearcore/blob/7ffb2698fc9fe4b908423f13b9352856cdaab26a/node/beacon/src/authority.rs#L307, which can be done with the current API but will make code less readable.

Reply:
I thought we want pruning to be domain-specific.

Reply:
The above pieces of code would change to:
```
self.storage.get_processed_slots(epoch)
```
```
if slot >0 && !self.storage.has_slot(&slot) {
```
```
self.storage.get_processed_slots(epoch).len()
```
(Btw, there is a typo, we should be passing epoch and not slot above)

Are you suggesting this is less readable? To me it looks more readable.

Reply:
It most cases pruning would be either time-based or LRU-based. We can have a domain-specific pruning strategy, but I do not think we have a strong need in it at the moment.